lib/kernel/src/code.erl

```diff
@@ -785,6 +834,97 @@ where_is_file(Tail, File, Path, Files) ->
             where_is_file(Tail, File)
     end.
 
+-spec get_doc(Mod) -> {ok, Res} | {error, Reason} when
+      Mod :: module(),
+      Res :: #docs_v1{},
+      Reason :: non_existing | missing | file:posix().
+get_doc(Mod) when is_atom(Mod) ->
+    case which(Mod) of
+        preloaded ->
+            Fn = filename:join([code:lib_dir(erts),"ebin",atom_to_list(Mod) ++ ".beam"]),
+            get_doc_chunk(Fn, Mod);
+        Error when is_atom(Error) ->
+            {error, Error};
+        Fn ->
+            get_doc_chunk(Fn, Mod)
+    end.
+
+get_doc_chunk(Filename, Mod) when is_atom(Mod) ->
+    case beam_lib:chunks(Filename, ["Docs"]) of
+        {error,beam_lib,{missing_chunk,_,_}} ->                
+            case get_doc_chunk(Filename, atom_to_list(Mod)) of
+                {error,missing} ->
+                    get_doc_chunk_from_ast(Filename);
+                Error ->
+                    Error
+            end;
+        {error,beam_lib,{file_error,Filename,enoent}} ->
+            get_doc_chunk(Filename, atom_to_list(Mod));
+        {ok, {Mod, [{"Docs",Bin}]}} ->
+            binary_to_term(Bin)
+    end;
+get_doc_chunk(Filename, Mod) ->
+    case filename:dirname(Filename) of
+        Filename ->
+            {error,missing};
+        Dir ->
+            ChunkFile = filename:join([Dir,"doc","chunks",Mod ++ ".chunk"]),
+            case file:read_file(ChunkFile) of
+                {ok, Bin} ->
+                    {ok, binary_to_term(Bin)};
+                {error,enoent} ->
+                    get_doc_chunk(Dir, Mod);
+                {error,Reason} ->
+                    {error,Reason}
+            end
+    end.
+
+get_doc_chunk_from_ast(Filename) ->
```

Comment:
I wonder if building a "fake chunk" from the AST on the fly is the way to go. Maybe `code:get_doc/1` should remain truthful to what is actually in the chunk/disk and returned an error in cases like this. What are the cases the "fake chunk" is necessary? :)

Reply:
I added it mostly because for modules without any chunk but still with a spec I thought it would be nice of `c:h` to actually return the spec.

This is mainly for erlang libraries without any API docs at all.

Reply:
I see. We do the same but the generation of the "fake chunk" is inside our "shell_docs". I guess some tools would find the "fake chunk" useful but others may not, so I guess both approaches work. Thanks for the info!